spec/lib/pact_broker/client/hal/http_client_spec.rb

```diff
@@ -157,6 +159,81 @@ module Hal
           end
         end
       end
+
+      describe "x509 certificate" do
+        FAKE_SERVER_URL = 'https://localhost:4444'
+        X509_CERT_FILE_PATH = './spec/fixtures/certificates/client_cert.pem'
+        X509_KEY_FILE_PATH = './spec/fixtures/certificates/key.pem'
+        UNSIGNED_X509_CERT_FILE_PATH = './spec/fixtures/certificates/unsigned_cert.pem'
+        UNSIGNED_X509_KEY_FILE_PATH = './spec/fixtures/certificates/unsigned_key.pem'
+
+        def wait_for_server_to_start
+          Faraday.new(
+            url: FAKE_SERVER_URL,
+            ssl: {
+              verify: false,
+              client_cert: OpenSSL::X509::Certificate.new(File.read(X509_CERT_FILE_PATH)),
+              client_key: OpenSSL::PKey::RSA.new(File.read(X509_KEY_FILE_PATH))
+            }
+          ) do |builder|
+            builder.request :retry, max: 20, interval: 0.5, exceptions: [StandardError]
+            builder.adapter :net_http
+          end.get
+        end
+
+        let(:do_get) { subject.get(FAKE_SERVER_URL) }
+
+        before(:all) do
+          @pipe = IO.popen("bundle exec ruby ./spec/support/ssl_server.rb")
+          ENV['SSL_CERT_FILE'] = "./spec/fixtures/certificates/ca_cert.pem"
+          ENV['SSL_CERT_DIR'] =  "./spec/fixtures/certificates/"
+
+          wait_for_server_to_start()
+        end
+
+        context "with valid x509 client certificates" do
+          before do
+            ENV['X509_CLIENT_CERT_FILE'] = X509_CERT_FILE_PATH
+            ENV['X509_CLIENT_KEY_FILE'] = X509_KEY_FILE_PATH
+          end
+
+          it "succeeds" do
+            expect(do_get.status).to eq 200
+          end
+        end
+
+        context "when invalid x509 certificates are set" do
+          before do
+            ENV['X509_CLIENT_CERT_FILE'] = UNSIGNED_X509_CERT_FILE_PATH
+            ENV['X509_CLIENT_KEY_FILE'] = UNSIGNED_X509_KEY_FILE_PATH
+          end
+
+          it "fails raising SSL error" do
+            expect { do_get }
+              .to raise_error { |error|
+                expect([OpenSSL::SSL::SSLError, Errno::ECONNRESET]).to include(error.class)
```

Comment:
That's curious. It could raise either error? Can you tell me more about that?

Reply:
Seems like so.
I ran the server in debug mode to check why this happening, and I got the following output by running it in:

<details>
  <summary>ubuntu-latest ruby v3.1</summary>
  
```
  [2023-09-20 08:52:01] INFO  WEBrick::HTTPServer#start: pid=508 port=4444
  |     with valid x509 client certificates
  |       succeeds
  |     when invalid x509 certificates are set
  | [2023-09-20 08:52:01] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 peeraddr=127.0.0.1:49104 state=error: certificate verify failed (self-signed certificate)
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
  | [2023-09-20 08:52:01] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 peeraddr=127.0.0.1:49106 state=error: certificate verify failed (self-signed certificate)
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
  | ERROR: Error making request - Errno::ECONNRESET Connection reset by peer /Users/jobandtalent/jobandtalent/pact_broker-client/lib/pact_broker/client/hal/http_client.rb:87:in `block (2 levels) in perform_request', attempt 1 of 5
  | [2023-09-20 08:52:06] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 peeraddr=127.0.0.1:49112 state=error: certificate verify failed (self-signed certificate)
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
  | [2023-09-20 08:52:06] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 peeraddr=127.0.0.1:49122 state=error: certificate verify failed (self-signed certificate)
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
  | ERROR: Error making request - Errno::ECONNRESET Connection reset by peer /Users/jobandtalent/jobandtalent/pact_broker-client/lib/pact_broker/client/hal/http_client.rb:87:in `block (2 levels) in perform_request', attempt 2 of 5
  | [2023-09-20 08:52:12] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 peeraddr=127.0.0.1:43724 state=error: certificate verify failed (self-signed certificate)
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
  | [2023-09-20 08:52:12] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 peeraddr=127.0.0.1:43736 state=error: certificate verify failed (self-signed certificate)
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
  | ERROR: Error making request - Errno::ECONNRESET Connection reset by peer /Users/jobandtalent/jobandtalent/pact_broker-client/lib/pact_broker/client/hal/http_client.rb:87:in `block (2 levels) in perform_request', attempt 3 of 5
  | [2023-09-20 08:52:17] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 peeraddr=127.0.0.1:43752 state=error: certificate verify failed (self-signed certificate)
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
  | [2023-09-20 08:52:17] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 peeraddr=127.0.0.1:43764 state=error: certificate verify failed (self-signed certificate)
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
  | ERROR: Error making request - Errno::ECONNRESET Connection reset by peer /Users/jobandtalent/jobandtalent/pact_broker-client/lib/pact_broker/client/hal/http_client.rb:87:in `block (2 levels) in perform_request', attempt 4 of 5
  | [2023-09-20 08:52:22] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 peeraddr=127.0.0.1:44812 state=error: certificate verify failed (self-signed certificate)
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
  | [2023-09-20 08:52:22] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 peeraddr=127.0.0.1:44818 state=error: certificate verify failed (self-signed certificate)
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
  |       /opt/hostedtoolcache/Ruby/3.1.4/x64/lib/ruby/gems/3.1.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
  | ERROR: Error making request - Errno::ECONNRESET Connection reset by peer /Users/jobandtalent/jobandtalent/pact_broker-client/lib/pact_broker/client/hal/http_client.rb:87:in `block (2 levels) in perform_request', attempt 5 of 5
  |       fails raising SSL error
  ```
</details>


<details>
  <summary>ubuntu-latest ruby v2.7</summary>
  
```
with valid x509 client certificates
|       succeeds
|     when invalid x509 certificates are set
| [2023-09-20 08:59:59] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 state=error: certificate verify failed (self signed certificate)
|       /opt/hostedtoolcache/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
|       /opt/hostedtoolcache/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
|       /opt/hostedtoolcache/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
|       /opt/hostedtoolcache/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
| [2023-09-20 08:59:59] ERROR OpenSSL::SSL::SSLError: SSL_accept returned=1 errno=0 state=error: certificate verify failed (self signed certificate)
|       /opt/hostedtoolcache/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `accept'
|       /opt/hostedtoolcache/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/webrick-1.8.1/lib/webrick/server.rb:302:in `block (2 levels) in start_thread'
|       /opt/hostedtoolcache/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/webrick-1.8.1/lib/webrick/utils.rb:258:in `timeout'
|       /opt/hostedtoolcache/Ruby/2.7.8/x64/lib/ruby/gems/2.7.0/gems/webrick-1.8.1/lib/webrick/server.rb:300:in `block in start_thread'
|       fails raising SSL error
  ```
</details>

Reply:
Seems like depending on the OS and ruby version it handles the verification error differently

https://github.com/ruby/net-http/blob/beb20c036dce1b14b8e78c012c3537957aaf6590/test/net/http/test_https.rb#L242

Reply:
Gotcha, thanks for digging into that.